Sources/NIOExtras/TaggedRequestResponseHandler.swift

```diff
@@ -0,0 +1,134 @@
+//
+//===----------------------------------------------------------------------===//
+
+import NIOCore
+
+/// `NIOTaggedRequestResponseHandler` receives a `Request` alongside an `EventLoopPromise<Response>` from the
+/// `Channel`'s outbound side. It will fulfill the promise with the `Response` once it's received from the `Channel`'s
+/// inbound side. Requests and responses can arrive out-of-order and are matched by the virtue of being
+/// `NIORequestIdentifiable`.
+///
+/// `NIOTaggedRequestResponseHandler` does support pipelining `Request`s and it will send them pipelined further down the
+/// `Channel`. Should `RequestResponseHandler` receive an error from the `Channel`, it will fail all promises meant for
+/// the outstanding `Reponse`s and close the `Channel`. All requests enqueued after an error occured will be immediately
+/// failed with the first error the channel received.
+///
+/// `NIOTaggedRequestResponseHandler` does _not_ require that the `Response`s arrive on `Channel` in the same order as
+/// the `Request`s were submitted. They are matched by their `requestID` property (from `NIORequestIdentifiable`).
+public final class NIOTaggedRequestResponseHandler<Request: NIORequestIdentifiable,
+                                                   Response: NIORequestIdentifiable>: ChannelDuplexHandler
+                                                  where Request.RequestID == Response.RequestID {
+    public typealias InboundIn = Response
+    public typealias InboundOut = Never
+    public typealias OutboundIn = (Request, EventLoopPromise<Response>)
+    public typealias OutboundOut = Request
+
+    private enum State {
+        case operational
+        case error(Error)
+
+        var isOperational: Bool {
+            switch self {
+            case .operational:
+                return true
+            case .error:
+                return false
+            }
+        }
+    }
+
+    private var state: State = .operational
+    private var promiseBuffer: [Request.RequestID: EventLoopPromise<Response>]
+
+
+    /// Create a new `RequestResponseHandler`.
+    ///
+    /// - parameters:
+    ///    - initialBufferCapacity: `RequestResponseHandler` saves the promises for all outstanding responses in a
+    ///          buffer. `initialBufferCapacity` is the initial capacity for this buffer. You usually do not need to set
+    ///          this parameter unless you intend to pipeline very deeply and don't want the buffer to resize.
+    public init(initialBufferCapacity: Int = 4) {
+        self.promiseBuffer = [:]
+        self.promiseBuffer.reserveCapacity(initialBufferCapacity)
+    }
+
+    public func channelInactive(context: ChannelHandlerContext) {
+        switch self.state {
+        case .error:
+            // We failed any outstanding promises when we entered the error state and will fail any
+            // new promises in write.
+            assert(self.promiseBuffer.count == 0)
+        case .operational:
```

Comment:
Do we need a state change here? I suspect the state machine is insufficiently large, because if a user _does_ `write` after we hit `channelInactive` then we'll buffer their promise and then leak it.

Reply:
Yes, thanks, fixed & tested